goalie/interpolation.py

```diff
@@ -1,72 +1,64 @@
-        seek to project into
-    :kwarg adjoint: apply the transposed projection operator?
+        seek to project into, or the :class:`firedrake.function.Function` or
+        :class:`firedrake.cofunction.Cofunction` to use as the target
     """
-    if not isinstance(source, firedrake.Function):
-        raise NotImplementedError("Can only currently project Functions.")  # TODO
-    Vs = source.function_space()
-    if isinstance(target_space, firedrake.Function):
+    if not isinstance(source, (firedrake.Function, firedrake.Cofunction)):
+        raise NotImplementedError(
+            "Can only currently project Functions and Cofunctions."
+        )
+
+    # If the input is a Cofunction then record this, map to a Function for the
+    # projection and then map back to a Cofunction afterwards
+    adjoint = isinstance(source, firedrake.Cofunction)
+    if adjoint:
+        source = cofunction2function(source)
+
+    # Account for cases where target_space is not a FunctionSpace
+    if isinstance(target_space, WithGeometry):
+        target = firedrake.Function(target_space)
+    elif isinstance(target_space, firedrake.Function):
         target = target_space
-        Vt = target.function_space()
+    elif isinstance(target_space, firedrake.Cofunction):
+        target = cofunction2function(target_space)
     else:
-        Vt = target_space
-        target = firedrake.Function(Vt)
-
-    # Account for the case where the meshes match
-    source_mesh = ufl.domain.extract_unique_domain(source)
-    target_mesh = ufl.domain.extract_unique_domain(target)
-    if source_mesh == target_mesh:
-        if Vs == Vt:
-            return target.assign(source)
-        elif not adjoint:
-            return target.project(source, **kwargs)
-
-    # Check validity of function spaces
-    space1, space2 = ("target", "source") if adjoint else ("source", "target")
-    if hasattr(Vs, "num_sub_spaces"):
-        if not hasattr(Vt, "num_sub_spaces"):
-            raise ValueError(
-                f"{space1} space has multiple components but {space2} space does not.".capitalize()
-            )
-        if Vs.num_sub_spaces() != Vt.num_sub_spaces():
-            raise ValueError(
-                f"Inconsistent numbers of components in {space1} and {space2} spaces:"
-                f" {Vs.num_sub_spaces()} vs. {Vt.num_sub_spaces()}."
-            )
-    elif hasattr(Vt, "num_sub_spaces"):
-        raise ValueError(
-            f"{space2} space has multiple components but {space1} space does not.".capitalize()
+        raise TypeError(
+            "Second argument must be a FunctionSpace, Function, or Cofunction."
         )
 
-    # Apply projector
-    return (_project_adjoint if adjoint else _project)(source, target, **kwargs)
+    # Choose appropriate transfer method
+    if source.function_space() == target.function_space():
+        target.assign(source)
+    elif adjoint:
+        target = _project_adjoint(source, target, **kwargs)
```

Comment:
Would it not make more sense to have `_projection_adjoint` only take a cofunctions as source and target (it only pulls out the vector anyway so it wouldn't need any conversion). I don't think doing an adjoint projection on normal functions really makes sense...

Reply:
Good point! Refactored in 1b825f0.